### PR TITLE
Upgrade couchdbkit

### PR DESCRIPTION
**Pull request by emord**

### CODE REVIEW

File: requirements/requirements.in [hunk 1]
```diff
@@ -2,8 +2,8 @@ attrs==18.1.0
 # used to build jsonobject from source
 cython>=0.29.1,<1.0.0
 iso8601==0.1.12
-jsonobject==0.9.4
-jsonobject-couchdbkit==0.9.12
+jsonobject>=0.9.4
+jsonobject-couchdbkit>=0.9.12
```
Comment (czue) [inline requirements/requirements.in]: should these be `~=0.9`?

**Review by czue — APPROVED** [review]

**Comment by czue** [comment]
kicked build

File: requirements/requirements.in (hunk 1 above)
Comment (emord) [reply requirements/requirements.in]: I also thought about that when doing this. I think my final thoughts were: any changes to these two libraries are intentionally made for HQ and there's not likely to be any API changes here that need to be thought about in HQ. I think my view of this file is:

* `==` this is either legacy, or we really need to pin this due to a bug or incompatibility in the library or we just don't trust the library to not introduce issues
* `~=` an external library that we want bugfix updates to and are fairly confident they follow semver
* `>=` internal libraries that we release for HQ and will never not want the latest version

That said I haven't actually run that by anyone. @dannyroberts you originally introduced pip-compile, how does that breakdown sound to you?

Comment (dannyroberts) [reply requirements/requirements.in]: Yeah that makes sense to me. Add to this the fact that (1) the pinned requirements files only get rebuilt when someone does so manually (2) there are separate commands for recompiling with as few changes as possible (`make requirements`) and for recompiling from scratch to get all latest versions (`make upgrade-requirements`), and (3) changes to the pinned requirements files go through code review, and I think there's enough options and human involvement for us to be able to notice and catch upgrades we shouldn't be doing. With that in mind, @emord I think your guidelines there give us a good starting point for sensible "suggested" upgrades when we proactively run `make upgrade-requirements`.

Thanks for articulating it like that, I don't think I ever laid it out so clearly when I originally introduced this.

Comment (czue) [reply requirements/requirements.in]: Nice that makes a lot of sense and sounds good. Is it something that should be documented/disseminated somewhere? I remember @dannyroberts gave a nice presentation on pip tools but not sure whether we created any lasting artifacts from that that would be good to have?

**Comment by snopoke** [comment]
@emord can you link to what the change were for this upgrade

**Comment by emord** [comment]
Change doc: https://github.com/cloudant/python-cloudant/blob/c296a5a5208d8e5850df939232c65713576e3855/CHANGES.md

Code changes: https://github.com/cloudant/python-cloudant/compare/2.7.0...2.10.1

**Comment by dannyroberts** [comment]
Thanks for that @emord. @snopoke I can confirm that it was for sure https://github.com/cloudant/python-cloudant/pull/412, which changed json parsing to go from `resp.json()` to `json.loads(resp.text)`. Sounds innocuous, right? Turns out that for an example doc on softlayer that I tried just now (the same one I was testing with last night), `resp.json()` is instantaneous, while `resp.text` takes 20+ seconds to return. That's because `resp.text` tries everything it can do to convert `str` to `unicode`, including using I believe a regular expression to try to infer what the proper encoding to use for decoding the bytes. It's my believe that it's that part—trying to infer the encoding—that takes a long time. Anyway, I can also confirm that `json.loads(resp.content)` (where content is the raw bytes (`str`) that `resp.text` and `resp.json()` work off of) is super fast and doesn't suffer from the same problem (and returns the same result as `json.loads(resp.text)`.

This is then a serious performance bug in this version of `python-cloudant`, and I think we should file an issue with them. The PR to fix is trivial (changing `resp.text` to `resp.content` in one location). I'll suggest this in the retro.

**Comment by emord** [comment]
Was the document the ICDS app doc? Could it be that `resp.text` is fast for small documents, but extremely slow for large ones? Its possible that we didn't see this on prod because the apps are much smaller than that doc

**Comment by dannyroberts** [comment]
Yeah it was definitely a very large doc (In domain reach-sandbox, so yeah some iteration of the ICDS app). I agree that whatever this was may have only really affected docs above a certain size.

**Comment by millerdev** [comment]
I think `json.loads(resp.content)` should be fine on Python 2 and Python 3.6+, but on older versions of Python 3 it may not work. Might want to make the decoding explicit like this:
```
json.loads(resp.content.decode('utf-8'))
```
[`json.loads()`](https://docs.python.org/3/library/json.html#json.loads) wants text, not bytes (until Python 3.6).

File: requirements/requirements.in (hunk 1 above)
Comment (dannyroberts) [reply requirements/requirements.in]: Hey @czue thanks for asking. When you asked the answer was no, but now thanks to your question the answer is yes :) Check out the new docs at https://confluence.dimagi.com/display/commcarehq/Python+Dependency+Versioning. Also linking them to https://confluence.dimagi.com/display/commcarehq/Home.

**Comment by dannyroberts** [comment]
Thank @millerdev, I'll include that in the issue/PR.

File: requirements/requirements.in (hunk 1 above)
Comment (czue) [reply requirements/requirements.in]: Nice, thanks! :raised_hands:

**Comment by dannyroberts** [comment]
@emord The issue in the python cloudant library has been fixed in `cloudant==2.10.2`: https://github.com/cloudant/python-cloudant/issues/414#issuecomment-448557702, so we're free to recreate this PR.

**Pull request merged**